test/affinity_tests.cc

```diff
@@ -0,0 +1,323 @@
+#include <celerity.h>
+
+#include <unordered_set>
+
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/matchers/catch_matchers.hpp>
+#include <catch2/matchers/catch_matchers_string.hpp>
+#include <libenvpp/env.hpp>
+#include <sycl/sycl.hpp>
+
+#include "affinity.h"
+#include "test_utils.h"
+
+using namespace celerity;
+
+namespace {
+
+using core_set = std::unordered_set<uint32_t>;
+
+// RAII utility for setting and restoring the process affinity mask
+class raii_affinity_masking {
+  public:
+	raii_affinity_masking(const core_set& cores, const bool clean_check = true) : m_clean_check(clean_check) {
+		REQUIRE(sched_getaffinity(0, sizeof(cpu_set_t), &m_prior_mask) == 0);
+		CPU_ZERO(&m_new_mask);
+		for(auto core : cores) {
+			CPU_SET(core, &m_new_mask);
+		}
+		REQUIRE(sched_setaffinity(0, sizeof(cpu_set_t), &m_new_mask) == 0);
+	}
+	~raii_affinity_masking() {
+		if(m_clean_check) {
+			cpu_set_t mask_after = {};
+			REQUIRE(sched_getaffinity(0, sizeof(cpu_set_t), &mask_after) == 0);
+			CHECK(CPU_EQUAL(&mask_after, &m_new_mask));
+		}
+		REQUIRE(sched_setaffinity(0, sizeof(cpu_set_t), &m_prior_mask) == 0);
+	}
+	raii_affinity_masking(const raii_affinity_masking&) = delete;
+	raii_affinity_masking(raii_affinity_masking&&) = delete;
+	raii_affinity_masking& operator=(const raii_affinity_masking&) = delete;
+	raii_affinity_masking& operator=(raii_affinity_masking&&) = delete;
+
+  private:
+	cpu_set_t m_prior_mask = {};
+	cpu_set_t m_new_mask = {};
+	bool m_clean_check = true;
+};
+
+// RAII utility to start the runtime with exactly `n` devices
+class raii_test_runtime {
+  public:
+	raii_test_runtime(int n) {
+		// devices are default constructible, and we don't care if we use the same more than once
+		std::vector<sycl::device> devices(n);
+		celerity::runtime::init(nullptr, nullptr, devices);
+	}
+	~raii_test_runtime() { celerity::runtime::shutdown(); }
+	raii_test_runtime(const raii_test_runtime&) = delete;
+	raii_test_runtime(raii_test_runtime&&) = delete;
+	raii_test_runtime& operator=(const raii_test_runtime&) = delete;
+	raii_test_runtime& operator=(raii_test_runtime&&) = delete;
+};
+
+constexpr void skip_unsupported_configurations() {
+#ifdef _WIN32
+	SKIP("Affinity is not supported on Windows");
+#endif
+}
+
+core_set get_current_cores() {
+	cpu_set_t mask = {};
+	REQUIRE(sched_getaffinity(0, sizeof(cpu_set_t), &mask) == 0);
+	core_set cores;
+	for(uint32_t i = 0; i < CPU_SETSIZE; ++i) {
+		if(CPU_ISSET(i, &mask)) { cores.insert(i); }
+	}
+	return cores;
+}
+
+bool have_cores(const core_set& desired_set) {
+	auto current_set = get_current_cores();
+	return std::ranges::all_of(desired_set, [&](uint32_t core) { return current_set.contains(core); });
+}
+
+} // namespace
+
+TEST_CASE("semantic thread type enum entries can be turned into strings", "[affinity]") {
+	using namespace detail::thread_pinning;
+	auto to_tt([&](uint32_t t) { return static_cast<thread_type>(t); });
```

Comment:
```suggestion
	const auto to_tt = [&](uint32_t t) { return static_cast<thread_type>(t); };
```
This is very odd to read.

Reply:
That is odd indeed, must have been late last Saturday.